File: lib/rgfa/line/gap/references.rb
# Sets the reference to the segments in gaps, when they are connected
# to a RGFA instance; creates virtual segments, if the segments have not
# been found yet.
#
# Computes the key to the reference to the gap in segments (gaps_R or gaps_L)
# depending on the orientations.
module RGFA::Line::Gap::References

  private

  def initialize_references
    [1,2].each do |snum|
      sid = :"sid#{snum}"
      orient = get(sid).orient
      linesymbol = get(sid).line
      s = @rgfa.segment(linesymbol)
      if s.nil?
        raise RGFA::NotFoundError if @rgfa.segments_first_order
        s = RGFA::Line::Segment::GFA2.new({:sid => linesymbol,
                                           :slen => 1,
                                           :sequence => "*"},
                                           version: :gfa2,
                                           virtual: true)
        s.connect(@rgfa)
      end
      set_existing_field(sid, OL[s,orient], set_reference: true)
      s.add_reference(self, refkey_for_s(snum))
    end
  end

  def refkey_for_s(snum)
    case [sid1.orient, sid2.orient]
    when [:+, :+]
      return (snum == 1) ? :gaps_R : :gaps_L
    when [:+, :-]
      return :gaps_R
    when [:-, :+]
      return :gaps_L
    when [:-, :-]
      return (snum == 1) ? :gaps_L : :gaps_R
    else
      raise RGFA::AssertionError, "Bug found, please report\n"+
        "snum: #{snum}"
    end
  end

end

File: gfapy/line/gap/references.py
import gfapy

class References:
  def _initialize_references(self):
    for snum in [1,2]:
      sid = "sid{}".format(snum)
      orient = self.get(sid).orient
      linesymbol = self.get(sid).line
      s = self._gfa.segment(linesymbol)
      if s is None:
        if self._gfa._segments_first_order:
          raise gfapy.NotFoundError()
        s = gfapy.line.segment.GFA2({"sid" : linesymbol,
                                     "slen" : 1,
                                     "sequence" : "*"},
                                    version = "gfa2",
                                    virtual = True)
        s.connect(self._gfa)
      self._set_existing_field(sid, gfapy.OrientedLine(s,orient),
          set_reference = True)
      s._add_reference(self, self._refkey_for_s(snum))

  def _refkey_for_s(self, snum):
    a = [self.sid1.orient, self.sid2.orient]
    if a == ["+", "+"]:
      return "gaps_R" if (snum == 1) else "gaps_L"
    elif a == ["+", "-"]:
      return "gaps_R"
    elif a == ["-", "+"]:
      return "gaps_L"
    elif a == ["-", "-"]:
      return "gaps_L" if (snum == 1) else "gaps_R"
    else:
      raise gfapy.AssertionError("Bug found, please report\n"+
                                 "snum: {}".format(snum))

  def _import_field_references(self, previous):
    for sid in ["sid1", "sid2"]:
      orient = self.get(sid).orient
      linesymbol = self.get(sid).line
      self._set_existing_field(sid,
          gfapy.OrientedLine(self._gfa.segment(linesymbol),orient),
          set_reference = True)
